jsonschema/src/output.rs

```diff
@@ -0,0 +1,281 @@
+        absolute_keyword_location: Option<AbsolutePath>,
+        annotations: Annotations<'_>,
+    ) -> OutputUnit<Annotations<'_>> {
+        OutputUnit {
+            keyword_location,
+            instance_location,
+            absolute_keyword_location,
+            value: annotations,
+        }
+    }
+
+    pub(crate) const fn error(
+        keyword_location: JSONPointer,
+        instance_location: JSONPointer,
+        absolute_keyword_location: Option<AbsolutePath>,
+        error: ErrorDescription,
+    ) -> OutputUnit<ErrorDescription> {
+        OutputUnit {
+            keyword_location,
+            instance_location,
+            absolute_keyword_location,
+            value: error,
+        }
+    }
+}
+
+#[derive(serde::Serialize, Debug, Clone, PartialEq)]
+pub struct Annotations<'a>(AnnotationsInner<'a>);
+
+#[derive(Debug, Clone, PartialEq)]
+enum AnnotationsInner<'a> {
+    UnmatchedKeywords(&'a AHashMap<String, serde_json::Value>),
+    ValueRef(&'a serde_json::Value),
+    Value(Box<serde_json::Value>),
+}
+
+impl<'a> From<&'a AHashMap<String, serde_json::Value>> for Annotations<'a> {
+    fn from(anns: &'a AHashMap<String, serde_json::Value>) -> Self {
+        Annotations(AnnotationsInner::UnmatchedKeywords(anns))
+    }
+}
+
+impl<'a> From<&'a serde_json::Value> for Annotations<'a> {
+    fn from(v: &'a serde_json::Value) -> Self {
+        Annotations(AnnotationsInner::ValueRef(v))
+    }
+}
+
+impl<'a> From<serde_json::Value> for Annotations<'a> {
+    fn from(v: serde_json::Value) -> Self {
+        Annotations(AnnotationsInner::Value(Box::new(v)))
+    }
+}
+
+#[derive(serde::Serialize, Debug, Clone, PartialEq)]
+pub struct ErrorDescription(String);
+
+impl From<ValidationError<'_>> for ErrorDescription {
+    fn from(e: ValidationError<'_>) -> Self {
+        ErrorDescription(e.to_string())
+    }
+}
+
+impl<'a> From<&'a str> for ErrorDescription {
+    fn from(s: &'a str) -> Self {
+        ErrorDescription(s.to_string())
+    }
+}
+
+impl<'a> serde::Serialize for BasicOutput<'a> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let mut map_ser = serializer.serialize_map(Some(2))?;
+        match self {
+            BasicOutput::Valid(outputs) => {
+                map_ser.serialize_entry("valid", &true)?;
+                map_ser.serialize_entry("annotations", outputs)?;
+            }
+            BasicOutput::Invalid(errors) => {
+                map_ser.serialize_entry("valid", &false)?;
+                map_ser.serialize_entry("errors", errors)?;
+            }
+        }
+        map_ser.end()
+    }
+}
+
+impl<'a> serde::Serialize for AnnotationsInner<'a> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        match self {
+            Self::UnmatchedKeywords(kvs) => {
+                let stdmap: std::collections::HashMap<String, serde_json::Value> =
```

Comment:
Would it work with `kvs` directly if we enable the `serde` feature from `ahash`?

Reply:
It does indeed!